kvdb-rocksdb: configurable memory budget per column

ordian opened the pull request:
Closes #135.
This is a breaking change (changes `DatabaseConfig`, makes `memory_budget_per_col` private, removed `write_rate_limit` option from `CompactionProfile`) .

ordian commented on kvdb-rocksdb/src/lib.rs:
```diff
@@ -286,8 +292,16 @@ impl Database {
 		opts.set_parsed_options("bytes_per_sync=1048576").map_err(other_io_err)?;
-		opts.set_db_write_buffer_size(config.memory_budget_per_col() / 2);
-		opts.increase_parallelism(cmp::max(1, ::num_cpus::get() as i32 / 2));
+		if config.memory_budget.is_empty() {
```
if we have at least one column, `write_buffer_size` is set automatically in `col_config` on `optimize_level_style_compaction`

grbIzl commented on kvdb-rocksdb/src/lib.rs:
```diff
@@ -152,47 +152,53 @@ impl CompactionProfile {
 	pub fn memory_budget(&self) -> usize {
-		self.memory_budget.unwrap_or(DB_DEFAULT_MEMORY_BUDGET_MB) * MB
+		if self.memory_budget.is_empty() {
+			return DB_DEFAULT_MEMORY_BUDGET_MB * MB;
```
If both fields are not set (memory_budget and memory_budget_per_col), all columns will have DB_DEFAULT_MEMORY_BUDGET_MB size and overall size will be the same DB_DEFAULT_MEMORY_BUDGET_MB. Is it ok?

ordian replied:
See the PR description:
> One of the things I'm not happy about in this PR, is that it doesn't allow to configure memory budget in case of 0 non-default columns (i.e. Database::open_default()), but it seems like we always use default settings for that case in parity-ethereum.

ordian replied:
I was thinking of making memory_budget an enum like
```rust
enum MemoryBudget {
    DefaultOnly(Option<usize>),
    PerColumn(Vec<Option<usize>>),
```
but that would complicate the API and we don't seem to use it anyway.

grbIzl replied:
Didn't understand properly your note in the description :-( 

Why not to use the current logic (memory_budget_for_col = total \ amount_of_cols), if exact values for columns are not set? It's a good heuristic. It's a just a total miss for Parity Ethereum. So in this case Parity Ethereum should use proper defaults for columns.

dvdplm replied:
> Why not to use the current logic (memory_budget_for_col = total \ amount_of_cols), if exact values for columns are not set?

I am also curious about this, can you elaborate @ordian ?

ordian replied:
(this was discussed privately, so I didn't reply)
The whole point of this change is to move the logic of dividing the the whole budget to the user. So e.g. parity-ethereum client will divide the budget and provide a config with `vec<90%, 10%, ... >` memory budget to kvdb-rocksdb. Does that make sense?

ordian replied:
If some parts are not set, it's not clear what the whole budget is how are we supposed to distribute it, so the fixed default amount is used.

dvdplm replied:
> So e.g. parity-ethereum client will divide the budget and provide a config with `vec<90%, 10%, ... >` memory budget to kvdb-rocksdb. Does that make sense?

I actually though the user were to provide absolute sizes; proportions make more sense. Docs definitely needed.

ordian replied:
No, I meant absolute sizes, sorry for confusion. Since the whole budget is not passed, it's the only info we have.

grbIzl commented:
I think, we need to make configurable the following things:
- keep_log_file_num (now it's hardcoded as 1 and it makes debug more difficult)
- block size (now it's set in Compaction Profile in accordance with disk's type). In  https://github.com/paritytech/parity-substrate-rocksdb-tuning was proposed to set block size to 8 Mb. Not it's hardcoded for ssd as 16Kb

grbIzl commented:
Relates to https://github.com/paritytech/parity-ethereum/issues/11233

ordian commented:
@grbIzl Added `keep_log_file_num` in 0ee93d1. Block size is already exposed e.g. via
```rust
CompactionProfile {
   block_size: 8 * 1024 * 1024,
   ..CompactionProfile::ssd(),
}
```

ordian commented on kvdb-rocksdb/src/lib.rs:
```diff
@@ -152,47 +152,56 @@ impl CompactionProfile {
 	pub compaction: CompactionProfile,
-	/// Set number of columns
-	pub columns: Option<u32>,
+	/// Specify the maximal number of info log files to be kept.
```
not sure if the docs are more elaborate https://github.com/facebook/rocksdb/blob/6ec6a4a9a49e506eff76aebd104d30be6a2d36cc/include/rocksdb/options.h#L595-L597

ordian replied:
AFAIK, an info log file is a `overlayrecent/db/LOG` file https://github.com/facebook/rocksdb/blob/9bd5fce6e89fcb294a1d193f32f3e4bb2e41d994/file/filename.cc#L187-L195

dvdplm replied:
Yeah, I think you're right. It's that log where developers go look for debug info (in an LSM, the word "log" is kind of inflated). Let's use `Specify the maximum number of info/debug log files to be kept.` – not elegant but clearer imo.

ordian commented:
> Not sure I understand this last part, can you elaborate?

Each time we add a new column in parity-ethereum, it kicks off a migration process, the budget for each column changes, which in turn may trigger compaction process that might take quite some time on archive nodes, so we would like to avoid that.

dvdplm commented on kvdb-rocksdb/src/lib.rs:
```diff
@@ -152,47 +152,56 @@ impl CompactionProfile {
-		self.memory_budget() / self.columns.unwrap_or(1) as usize
+	/// # Panics
+	///
+	/// If `col` is out of bounds.
```
```suggestion
	/// Will panic if `col` is greater than the length of `self.memory_budget`.
```

dvdplm commented on kvdb-rocksdb/src/lib.rs:
```diff
@@ -642,21 +659,17 @@ impl Database {
 
 	/// Drop a column family.
-	pub fn drop_column(&self) -> io::Result<()> {
+	pub fn drop_column(&mut self) -> io::Result<()> {
```
So we can only drop columns in last-first order? That's an odd API isn't it?

ordian replied:
yes, it is, but I'm not even sure if we've ever used it

dvdplm replied:
Can we get rid of it?

ordian replied:
I mean technically we use it here https://github.com/paritytech/parity-ethereum/blob/db1ea1dcd8361c6b53a136842398464da363447b/util/migration-rocksdb/src/lib.rs#L304, but in practice, we never dropped columns on db version bump.

dvdplm commented on kvdb-rocksdb/src/lib.rs:
```diff
@@ -642,21 +659,17 @@ impl Database {
 				if let Some(col) = cfs.pop() {
 					let name = format!("col{}", cfs.len());
 					drop(col);
+					self.config.memory_budget.resize(cfs.len(), None);
```
So this could even have been `drop(self.config.memory_budget.pop())`?

dvdplm commented on kvdb-rocksdb/src/lib.rs:
```diff
@@ -666,12 +679,17 @@ impl Database {
+					self.config.memory_budget.resize(col + 1, None);
+				}
+				let memory_budget_per_col = self.config.memory_budget_per_col(col);
+				let col_config = col_config(&self.config, &self.block_opts, memory_budget_per_col)?;
```
Couldn't `col_config()` be a method of `DatabaseConfig` and have access to it's own config and be able to do all the setup there? Same for line [335](https://github.com/paritytech/parity-common/pull/256/files#diff-4382bfb7e75959b2b8884268540abfa3R335)?

dvdplm commented on kvdb-rocksdb/src/lib.rs:
```diff
@@ -152,47 +152,56 @@ impl CompactionProfile {
+	///
+	/// If `col` is out of bounds.
+	pub fn memory_budget_per_col(&self, col: usize) -> usize {
+		self.memory_budget[col].unwrap_or(DB_DEFAULT_MEMORY_BUDGET_MB) * MB
```
Let's do the multiplication with `MB` in the const definition.

ordian replied:
Not sure what you mean. Create a const function that multiples `usize` param by `MB`? What would be the benefit?

dvdplm replied:
No, I think I misread this, thought we did `DB_DEFAULT_MEMORY_BUDGET_MB * MB` in a few places and suggested we just do the multiplication when defining the const: `const DB_DEFAULT_MEMORY_BUDGET_MB: usize = 128 * MB;` but then we'd have to pass in the memory budget in bytes.
…so we need docs here to remind readers that the return value is in bytes. (And elsewhere that the `memory_budget` items are in mb).

dvdplm commented on kvdb-rocksdb/src/lib.rs:
```diff
@@ -152,47 +152,56 @@ impl CompactionProfile {
 impl DatabaseConfig {
 	/// Create new `DatabaseConfig` with default parameters and specified set of columns.
 	/// Note that cache sizes must be explicitly set.
 	pub fn with_columns(columns: Option<u32>) -> Self {
```
Wouldn't `new_with_columns` be a more idiomatic name for a constructor?

dvdplm commented on kvdb-rocksdb/src/lib.rs:
```diff
@@ -152,47 +152,56 @@ impl CompactionProfile {
 	pub compaction: CompactionProfile,
-	/// Set number of columns
-	pub columns: Option<u32>,
+	/// Specify the maximal number of info log files to be kept.
```
```suggestion
	/// Specify the maximum number of info log files to be kept.
```

ordian commented on kvdb-rocksdb/src/lib.rs:
```diff
@@ -284,10 +293,18 @@ impl Database {
+			// L1 size is controlled by the option `max_bytes_for_level_base`.
+			opts.set_parsed_options(&format!("max_bytes_for_level_base={}", budget)).map_err(other_io_err)?;
+		}
+		opts.increase_parallelism(cmp::max(1, num_cpus::get() as i32 / 2));
```
Maybe you're right, but that needs investigation and out of scope of this PR.

dvdplm replied:
[Looks like they default to 16 threads](https://github.com/facebook/rocksdb/blob/6ec6a4a9a49e506eff76aebd104d30be6a2d36cc/include/rocksdb/options.h#L362-L367) in the rocksdb source so I am probably completely wrong. Needs investigation.

ordian commented on kvdb-rocksdb/src/lib.rs:
```diff
@@ -152,47 +152,56 @@ impl CompactionProfile {
 }
 
 /// Database iterator (for flushed data only)
 // The compromise of holding only a virtual borrow vs. holding a lock on the
```
I've changed the logic here in #257.

ordian commented:
@dvdplm I've changed the impl to keep the `columns` field and added a `HashMap<Column, MemoryBudget>`. `indexmap` doesn't really help here, since it indexes based on insertion order, not the actual keys.

ordian commented:
> The idea here is to make minimal changes to leave the old API as intact as possible to avoid breakage for consuming code?

Yes, but I'm open to ideas how to make the API more ergonomic while keeping the migration overhead low.

ordian commented on kvdb-rocksdb/src/lib.rs:
```diff
@@ -152,47 +161,80 @@ impl CompactionProfile {
+		opts.optimize_level_style_compaction(memory_budget_per_col as i32);
+		opts.set_target_file_size_base(self.compaction.initial_file_size);
+
+		opts.set_parsed_options("compression_per_level=").map_err(other_io_err)?;
```
This was added in https://github.com/paritytech/parity-ethereum/pull/7348/commits/3cf1aac5b9e8fe23c68e588ee1a74304a7fecf5c by @andresilva with the intention of enabling compression on all levels, otherwise optimize_level_style_compaction [will disable](https://github.com/facebook/rocksdb/blob/6ec6a4a9a49e506eff76aebd104d30be6a2d36cc/options/options.cc#L548-L550) compression for the first two levels IIUC. #257 uses https://docs.rs/rocksdb/0.13.0/rocksdb/struct.Options.html#method.set_compression_per_level instead of `set_parsed_option`.

ordian commented:
I've updated the default block size for `ssd()` `CompactionProfile` to `8 MiB` and removed `write_rate_limit` in https://github.com/paritytech/parity-common/pull/256/commits/130ca3ef20b10f887f05b194beff8529fada8c20.

ordian commented on kvdb-rocksdb/src/lib.rs:
```diff
@@ -284,10 +293,18 @@ impl Database {
+			// Memtable size is controlled by the option `write_buffer_size`.
+			// If you increase your memtable size, be sure to also increase your L1 size! 
+			// L1 size is controlled by the option `max_bytes_for_level_base`.
+			opts.set_parsed_options(&format!("max_bytes_for_level_base={}", budget)).map_err(other_io_err)?;
```
I've set it to the same value as `write_buffer_size`, couldn't find any advice on that.
